system-parachains/asset-hubs/asset-hub-polkadot/src/staking/mod.rs

```diff
@@ -252,6 +254,140 @@ impl multi_block::unsigned::miner::MinerConfig for Runtime {
+// To be replaced after that is merged and available.
+pub mod temp_curve {
+	use super::*;
+	use sp_runtime::traits::{AtLeast32BitUnsigned, One};
+	use scale_info::TypeInfo;
+	use sp_arithmetic::traits::Saturating;
+
+	/// The step type for the stepped curve.
+	#[derive(PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo, Clone)]
+	pub enum Step<V> {
+		/// Move towards a desired value by a percentage of the remaining difference at each step.
+		///
+		/// Step size will be (target_total - current_value) * pct.
+		RemainingPct(V, Perbill),
+	}
+
+	/// A stepped curve.
+	///
+	/// Steps every `period` from the `initial_value` as defined by `step`.
+	/// First step from `initial_value` takes place at `start` + `period`.
+	#[derive(PartialEq, Eq, sp_core::RuntimeDebug, TypeInfo, Clone)]
+	pub struct SteppedCurve<P, V> {
+		/// The starting point for the curve.
+		pub start: P,
+		/// An optional point at which the curve ends. If `None`, the curve continues indefinitely.
+		pub end: Option<P>,
+		/// The initial value of the curve at the `start` point.
+		pub initial_value: V,
+		/// The change to apply at the end of each `period`.
+		pub step: Step<V>,
+		/// The duration of each step.
+		pub period: P,
+	}
+
+	impl<P, V> SteppedCurve<P, V>
+	where
+		P: AtLeast32BitUnsigned + Copy,
+		V: AtLeast32BitUnsigned + Copy + From<P>,
+	{
+		/// Creates a new `SteppedCurve`.
+		pub fn new(start: P, end: Option<P>, initial_value: V, step: Step<V>, period: P) -> Self {
+			Self { start, end, initial_value, step, period }
+		}
+
+		/// Returns the magnitude of the step size occuring at the start of this point's period.
+		/// If no step has occured, will return 0.
+		///
+		/// Ex. In period 4, the last step taken was 10 -> 7, it would return 3.
+		pub fn last_step_size(&self, point: P) -> V {
+			// No step taken yet.
+			if point < self.start {
+				return V::zero();
+			}
+
+			// If the period is zero, the value never changes.
+			if self.period.is_zero() {
+				return V::zero();
+			}
+
+			// Determine the effective point for calculation, capped by the end point if it exists.
+			let _effective_point = self.end.map_or(point, |e| point.min(e));
+
+			// Calculate how many full periods have passed.
+			let num_periods = (point - self.start) / self.period;
+
+			if num_periods.is_zero() {
+				return V::zero();
+			}
+
+			// Points for calculating step difference.
+			let prev_period_point = self.start + (num_periods - P::one()) * self.period;
+			let curr_period_point = self.start + num_periods * self.period;
+
+			// Evaluate the curve at those two points.
+			let val_prev = self.evaluate(prev_period_point);
+			let val_curr = self.evaluate(curr_period_point);
+
+			if val_curr >= val_prev {
+				return val_curr.saturating_sub(val_prev);
+			} else {
+				return val_prev.saturating_sub(val_curr);
+			}
+		}
+
+		/// Evaluate the curve at a given point.
+		pub fn evaluate(&self, point: P) -> V {
```

Comment:
```suggestion
		pub fn evaluate(&self, num_periods: ...) -> V {
```
Given that this is a discrete function, i think it could make sense to keep it that way and call it directly with discrete `num_periods` instead of up- and down-sampling the point here.